DDS Support for Arduino Zero (and its built-in DAC)

jake-b opened the pull request:
This is a very rough draft of DDS support for the Arduino Zero.  I still have more work to do, but it is working pretty well.   You can see the oscilloscope output from the code running on a Zero here:
- https://github.com/jake-b/HamShield/blob/assets/hardcopy_600.png
- https://github.com/jake-b/HamShield/blob/assets/hardcopy-2400.png

A nice smooth wave (more or less) without any filtering, since it is a DAC rather than PWM.

There are some limiting factors to using HamShield (as best I can gather from public info):
- The Arduino Zero is a 3.3v board and is not 5v compatible.  For proper support of newer Arduino boards, HamShield would need to include level shifters and use the IOREF pin to set the voltage on the processor side to support both 3.3v and 5v main boards.
- The HamShield expects the PWM audio out from the Arduino to be on D3.  The Zero's fancy 10-bit DAC is on A0.  (a selectable solder jumper maybe?)
- I2C for modern Arduino boards (Leonardo+) is on the SDA/SCL pins near D13.  You can bit-bang A4/A5 on the Zero if necessary, and I've tested it as working.

I'm going to try a "DigiX level shifter shield" between the HamShield and the Zero to deal with the 3.3v issue. The shield also lets you route the SDA/SCL lines to the A4/A5 location.   I can probably cut/solder around the A0/D3 issue as well. 

This is all out-of-scope for the Kickstarter, but I would love to see a "HamShield 2.0" that is more compatible with Leonardo, Due, and Zero.

I'll continue to work on this Zero DDS code to fix a few omissions and bugs.

jake-b committed:
Very rough draft of DDS supporting the Arduio Zero and it's built-in DAC.

jake-b committed:
- Fixed a few problems with Arduino Zero support
- Added a 10-bit sine table, probably wasteful considering its stored in 16-bit integers.
- Added "DDS_MAX_COMPARATOR" and "DDS_MAX_ACCUMULATOR" macros to hide some math and imrpove readability.
- Simplified calcFrequency and made it work for arbirary frequencies (with/without default refclk)
- Simplified getDutyCycle().  I think its more understandable now, and works for various combinations of accumulator size and lookup table size.

Tested on both an Arduino Zero and an Arduino UNO.  Seems to work, but YMMV.

jake-b commented:
I'm considering this "done" for now, until I get a real HamShield to play with.   Even if you don't take these changes, there are a few good ideas (I think) that you may want to incorporate into your main branch.

spaceneedle commented:
Thanks! This is really cool! I will have Stephen and Nigel take a look at
this.

We can get any of the bugs sorted out once you get your HamShield.

On Mon, Aug 17, 2015 at 2:31 PM, jake-b notifications@github.com wrote:

> I'm considering this "done" for now, until I get a real HamShield to play
> with. Even if you don't take these changes, there are a few good ideas (I
> think) that you may want to incorporate into your main branch.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/EnhancedRadioDevices/HamShield/pull/15#issuecomment-131968298
> .

jake-b committed:
Updated examples/DDS/DDS.ino to include the ISRs for the Arduino Zero.
NOTE: this sample won't really work for the HamShield ont he Ardiuno Zero as it will
Besides the logic voltage issue, it will be playing its tones out of AO, rather
than D3 or D11.  Hook up your oscilloscope to A0 to see the waveform.

spaceneedle commented:
Hello Jake,
could you send me your contact information? spaceneedle@gmail.com. I wanted to touch base with you on getting this working on the Zero. We had a few other folks who were interested!

aperezbios commented:
@spaceneedle, what's the current plan for merging this commit to support the Zero? Have you or anyone else validated it still works with the shipping version of HamShield?

nigelvh commented:
Has anyone had a chance to test this? I don't have access to a Zero board, so I'm unable to verify functionality.
